File: python/04_Binary_Search/binary_search.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
            # create left and right pointers
        l, r = 0, len(nums) - 1

        # iterate through the pointers
        while l <= r:
            # calculate the midpoint using floor division
            m = l + ((r - l) // 2)  # (l + r) // 2 can lead to overflow

            # if num higher than midpoint change right pointer
            if nums[m] > target:
                # the right pointer now updated
                r = m - 1
            
            # if num lower than midpoint update left pointer 
            elif nums[m] < target:
                l = m + 1

            # otherwise the num is found in mid point
            else:
                return m    # return index
        
        # return -1 if all fails 
        return -1

# implementation
answer = Solution()
nums = [-1, 0, 3, 5, 9, 12]

print("Leetcode 704: Binary Search\n")

# Example 1
target_1 = 9
print(f"Example 1: nums = {nums}, target = {target_1}")
print("Output:", answer.search(nums, target_1))

# Example 2
tasrget_2 = 2
print(f"Example 2: nums = {nums}, target = {tasrget_2}")
print("Output:", answer.search(nums, tasrget_2))

File: python/04_Binary_Search/binarySearch.go
package main

import "fmt"

// Function to perform binary search on a sorted array
func search(nums []int, target int) int {
	left := 0              // Pointer for the leftmost index of the search range
	right := len(nums) - 1 // Pointer for the rightmost index of the search range

	// Continue searching as long as the left pointer doesn't surpass the right pointer
	for left <= right {
		middle := (left + right) / 2 // Calculate the middle index of the current search range

		// Check if the middle element is equal to the target
		if nums[middle] == target {
			return middle // Return the index where the target is found
		} else if nums[middle] < target {
			left = middle + 1 // If the target is greater than the middle element, search the right half
		} else {
			right = middle - 1 // If the target is less than the middle element, search the left half
		}
	}
	return -1 // Return -1 if the target is not found in the list
}

// implementation
func main() {
	fmt.Println("Leetcode 704: Binary Search")
	nums := []int{-1, 0, 3, 5, 9, 12}

	// Example 1
	target1 := 9
	fmt.Printf("Example 1: nums = %v, target = %v\n", nums, target1)
	fmt.Println("Output:", search(nums, target1))

	// Example 2
	target2 := 2
	fmt.Printf("Example 2: nums = %v, target = %v\n", nums, target2)
	fmt.Println("Output:", search(nums, target2))
}
